woocommerce/woocommerce-ios
# [Bundle products in order form] UI updates on Configure screen and bundled order item rows

Part of #10428 
Based on https://github.com/woocommerce/woocommerce-ios/pull/11018
⚠️ Might conflict with https://github.com/woocommerce/woocommerce-ios/pull/11034


## Why


We recently got some design feedback on the Configure screen from Joe in pe5pgL-3Vt-p2#comment-3172. Plus, we want to update the design for bundled order items in the order form product list. We don't have design review on that yet, as an initial design improvement I decided to remove the border and set a tertiary system background color to the background for the bundled child order items.

## How

### Configure screen

Most of the changes are in `ConfigurableBundleItemView` and `ConfigurableBundleItemViewModel`.
- The checkbox is now always shown
- The checkbox is pre-selected and disabled if the bundle item is not optional
- If the bundle item is optional, the quantity changer is only visible when the item is selected
  - For this change, `ProductRowViewModel` in the view model has to become an Published observable. After it's initialized, `isOptionalAndSelected` is observed to create a corresponding `ProductRowViewModel` with `canChangeQuantity` based on the latest value

### Bundled child order item row UI

In order to differentiate the UI when a product row is a child item, a new bool property is added as `ProductRowViewModel.hasParentProduct`. There are two UI changes between the parent vs. child product rows:

- Background color: for child product rows, the background color is tertiary system background color
  - `CollapsibleView` used to have a fixed color, this PR updated so that a custom background color can be set
- In `CollapsibleProductRowCard `, the border of the product row card isn't shown for child product rows

For bundled child order items, the quantity changer UI is now hidden since the bundled item quantity should be configured in the parent bundle product. A test case was added for this.

## Testing instructions


Prerequisite: the store has [Product Bundles extension](https://woocommerce.com/products/product-bundles/) enabled, and also a bundle product with non-empty items. **The store has an order with a bundle product with child items.**

- Go to the Orders tab
- Tap on the order in the prerequisite (with a bundle product and child items)
- Tap `Edit` to edit the order --> rows should be shown as the design in the screenshots
- Tap on the expand  🔽 CTA on the **bundled child order item** --> the quantity changer shouldn't be shown
- Tap on the expand  🔽 CTA on the **bundle product order item** , then tap `Configure` --> after a bit (products being loaded remotely), the configurable bundle items should be shown. The checkbox should be shown for all optional and non-optional items
- Unselect an optional item if possible --> the quantity changer should be hidden
- Select an optional item if possible --> the quantity changer should be shown

## Screenshots


### Configure UI

optional - unselected | optional - selected
-- | --
<img src="https://github.com/woocommerce/woocommerce-ios/assets/1945542/757f56af-4ce4-4b6b-92cd-50412098acb4" width="300" /> | <img src="https://github.com/woocommerce/woocommerce-ios/assets/1945542/88f326ad-415e-4120-9601-baa3766284ce" width="300" />

### Order form product rows

dark | light
-- | --
<img src="https://github.com/woocommerce/woocommerce-ios/assets/1945542/c95224d8-f22d-470d-962e-c034a486bebc" width="300" /> | <img src="https://github.com/woocommerce/woocommerce-ios/assets/1945542/bdaab372-9a90-4f4e-897e-5428c3042302" width="300" />



---
